src/pdf_parser/pdf_utils/ocr.py

```diff
@@ -1,11 +1,329 @@
+    coordinates: BoundingPoly
+    confidence: float
+    language: Optional[str]
+
+
+class GoogleBlock(BaseModel):
+    """A fully structured block from google OCR. Can contain multiple segments."""
+
+    coordinates: BoundingPoly
+    text_blocks: List[GoogleTextSegment]
+
+
+def image_bytes(image: PpmImageFile) -> bytes:
+    """Return the image as a bytes object."""
+    image_buffer = io.BytesIO()
+    image.save(image_buffer, format="PNG")
+    image_buffer.seek(0)
+    return image_buffer.read()
+
+
+def google_vertex_to_shapely(bound):
+    return Polygon(
+        [
+            (bound.vertices[0].x, bound.vertices[0].y),
+            (bound.vertices[1].x, bound.vertices[1].y),
+            (bound.vertices[2].x, bound.vertices[2].y),
+            (bound.vertices[3].x, bound.vertices[3].y),
+        ]
+    )
+
+
+def extract_google_layout(
+    image: PpmImageFile,
+) -> Tuple[
+    List[GoogleBlock],
+    List[GoogleTextSegment],
+    List[GoogleTextSegment],
+    List[GoogleTextSegment],
+]:
+    """Returns document bounds given an image.
+
+        The Google OCR API returns blocks of paragraphs. Roughly, there are 3 cases worth considering:
+
+    1. A block consists of a heading and a paragraph of text below it. In this case the block consists of 2
+    paragraphs, the first being the heading and the second being the text.
+    2. A block consists of a paragraph of
+    text. In this case the block and the paragraph have the same coordinates.
+    3. A block consists of multiple  paragraphs of text. For example, if we have a list of items,
+    each item is a paragraph.
+
+    Given this, and given that we want to cross-reference the returned layout with the one returned by
+    layoutparser + heuristics, we store blocks and paragraphs separately as "GoogleBlocks". To see the utility of this,
+    consider 2 cases:
+
+    1. Layout parser classifies a heading and a paragraph of text below it as 2 separate blocks,
+     one for a heading and another for the following paragraph. In this case, the best coordinate matches
+      with LayoutParser will be on what google classifies as paragraphs.
+    2. Layout parser detects a list of items as a single block but is unable to detect the individual items.
+    In this case, the best matches will be on what google classifies as blocks. So we want to merge the text
+    from the paragraphs into the blocks.
+
+    Regarding point 2, Google appears to be able to detect the individual items in a list more reliably than
+    layoutparser, a point we can explore in the future to get the best results by combining the two approaches.
+    For future usability, we also store block objects with paragraphs as sub-objects.
+
+    Args:
+        image: Image to extract document bounds from.
+
+    Returns:
+        List of GoogleBlocks, List of GoogleTextSegments, List of GoogleTextSegments, List of GoogleTextSegments
+    """
+
+    def _get_modal_string(string_list: List[str]) -> Optional[str]:
```

Comment:
Nested functions are hard to test etc.?

Reply:
ok moved oout and added docstring to make clearer